vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -497,27 +523,215 @@ static void run_dijkstra(int start_node_ind,
+                            cache->insert(compressed_key, val.cost_entry.delay);
+                        }
+                    }
+                }
+
+                parent_entry = util::PQ_Entry(*it, parent_node.edge_switch(paths[*it].edge), parent_entry.delay,
+                                              parent_entry.R_upstream, parent_entry.congestion_upstream, false);
+                parent = *it;
+            }
+        } else {
+            expand_dijkstra_neighbours(current, paths, node_expanded, pq);
+            node_expanded[node_ind] = true;
         }
 
-        expand_dijkstra_neighbours(current, node_visited_costs, node_expanded, pq);
-        node_expanded[node_ind] = true;
+        max_cost = std::max(max_cost, current.delay_cost);
+        if (max_cost > cost_limit) {
+            break;
+        }
     }
+    return max_cost;
 }
 
-void ConnectionBoxMapLookahead::compute(const std::vector<t_segment_inf>& segment_inf, const std::string& lookahead_search_locations) {
-    compute_connection_box_lookahead(segment_inf, lookahead_search_locations);
+// compute the cost maps for lookahead
+void ConnectionBoxMapLookahead::compute(const std::vector<t_segment_inf>& segment_inf) {
+    vtr::ScopedStartFinishTimer timer("Computing connection box lookahead map");
+
+    size_t num_segments = segment_inf.size();
+    std::vector<SamplePoint> sample_points;
+    {
+        std::vector<SampleGrid> inodes_for_segment(num_segments);
+        find_inodes_for_segment_types(&inodes_for_segment);
+
+        // collapse into a vector
+        for (auto& grid : inodes_for_segment) {
+            for (int y = 0; y < SAMPLE_GRID_SIZE; y++) {
+                for (int x = 0; x < SAMPLE_GRID_SIZE; x++) {
+                    auto& point = grid.point[y][x];
+                    if (!point.samples.empty()) {
+                        point.order = point.samples[0];
+                        sample_points.push_back(point);
+                    }
+                }
+            }
+        }
+    }
+
+    // sort by VPR coordinate
+    std::sort(sample_points.begin(), sample_points.end(),
+              [](const SamplePoint& a, const SamplePoint& b) {
+                  return a.order < b.order;
+              });
+
+    /* free previous delay map and allocate new one */
+    auto& device_ctx = g_vpr_ctx.device();
+    cost_map_.set_counts(segment_inf.size(),
+                         device_ctx.connection_boxes.num_connection_box_types());
+
+    VTR_ASSERT(REPRESENTATIVE_ENTRY_METHOD == SMALLEST);
+    RoutingCosts all_costs;
+
+    /* run Dijkstra's algorithm for each segment type & channel type combination */
+#if defined(VPR_USE_TBB)
+    tbb::mutex all_costs_mutex;
+    tbb::parallel_for_each(sample_points, [&](const SamplePoint& point) {
+#else
+    for (const auto& point : sample_points) {
+#endif
+        float max_cost = 0.f;
+        RoutingCosts costs;
+        SimpleCache<CompressedRoutingCostKey, float, DIJKSTRA_CACHE_SIZE> cache;
```

Comment:
Would be good to explain what "SimpleCache" is doing

Reply:
I've added some comments to explain,